### DISCUSSION TITLE
Session management and profiles

### DISCUSSION THREAD

**Discussion by elviskahoro** [warpdotdev/Warp#697] [Product roadmap]
Placeholder 1

![A](https://user-images.githubusercontent.com/29553206/151546532-fa0947e0-c8ff-4039-8fb7-fd53b443d138.png)

![B](https://user-images.githubusercontent.com/29553206/151546560-da220bb5-ec9d-4def-a892-e82d8e5e844b.png)

![C](https://user-images.githubusercontent.com/29553206/151546602-67fa7bcc-af6d-4f35-b2b3-aa32e9bf5fa2.png)

![C](https://user-images.githubusercontent.com/29553206/151546626-2291f2ad-193d-4871-9d9a-3b11219970cd.png)

**Comment by elviskahoro** [comment]
Place holder 2

**Comment by rencryptofish** [comment]
it might be hard to tell which panes in a layout correspond to panes in the list and the startup commands, what about like a faint numbering that associates the panes in the layout with the pane's startup/dir

**Comment by elviskahoro** [comment]
From gtx via discord: https://discord.com/channels/851854972600451112/852249738002956311/940508883095552000

> The ability to use different themes to different windows (or tabs).
> 
> For certain tasks I always open a different new terminal window, and I would like it to be distinct so I could (a) quickly identify it (by color), (b) remember that I'm on a session with possible non-trivial side effects / important task.
>

**Reply by guytavor** [reply]
Thanks @elviskahoro . This being part of a "profile" (like it is in iterm2) makes lots of sense to me.

**Comment by elviskahoro** [comment]
It be nice if one could duplicate tabs or if there are tab presets #718

From @[ppwfx](https://github.com/ppwfx)

### Describe the solution you'd like?

```shell
It be nice if one could duplicate tabs or if there are tab presets
```


### Is your feature request related to a problem? Please describe.

_No response_

### Additional context

_No response_

**Comment by elviskahoro** [comment]
From [SandyChapman](https://github.com/SandyChapman) via #190 comment:
https://github.com/warpdotdev/Warp/issues/190#issuecomment-1079081593

> @nerdalytics Yeah, but there's also this setting in Terminal which has the same effect which also happens to be Apple's [recommended way to set the shell in Terminal](https://support.apple.com/en-ca/guide/terminal/trml113/mac). It's also mentioned here as [an alternative](https://support.apple.com/en-us/HT208050) to setting your login shell. And [there are reasons](https://superuser.com/a/1316833/275543) why using a non-posix login shell could be a bad idea.
> 
> ![image](https://user-images.githubusercontent.com/3920249/160139164-81fbded3-ae00-499f-8e89-3bf5779c3178.png)
>

**Reply by elviskahoro** [reply]
@SandyChapman @nerdalytics @gertjana

**Reply by SandyChapman** [reply]
Thanks @elviskahoro !

**Comment by elviskahoro** [comment]
> I use tmuxinator for restoring workspaces and for automatically running commands such as docker-compose start. I would like to have this natively supported by warp so I don’t have to be in tmux all day.

**Comment by dimfeld** [comment]
I originally posted this on [the TMUX topic](https://github.com/warpdotdev/Warp/discussions/501) but it probably makes more sense here.

What I really want (for my personal needs) isn't necessarily "actual tmux integrated with Warp" but the ability to detach from a window while keeping its tabs/panes and processes running in the background, and then later bring it back. 

I can kind of do this by minimizing windows but then it's difficult to know which is which just by looking at the minimized windows in the dock. I think for my needs, the following would be great:

- Give a name to a window (akin to tmux session names)
- Hide the current Warp window while leaving other Warp windows visible, like `tmux detach`.
- Bring windows back using the command palette or some similar fast way
- Ideally, a way to both detach the current window and switch to another with a single command, like the `tmux switch-client` command.

If Warp did this natively I wouldn't care at all that it's not real tmux.

**Reply by felixaa** [reply]
I completely agree! If it's complicated to include the "Warp shell" into tmux, why not build the "The warp way" of managing sessions.
I think you guys would reach a wider audience of developers that uses a CLI-based workflow.

I'd love to give your product a shot! But the lack of tmux or a session manager is a deal breaker

What I'd love is to have a keyboard driven approach with the ability to have VIM esc bindings to move around and manage this

-  The ability to switch between windows and panes and also renaming them

- Listing and switching between sessions and renaming them

- Minimalist look and customisable keybindings

**Comment by elviskahoro** [comment]
Need to checkout: https://github.com/eggplanetio/mert

**Comment by ggaabe** [comment]
It would be amazing if Warp could let people build workspaces in certain folders (similar to .vscode workspaces), and then let them build custom commands for those workspaces. On top of that, let them build a GUI-based command center, with a grid of large buttons with customizable icons and colors, where the user can set which commands each button triggers.

**Comment by rgaino** [comment]
> From gtx via discord: https://discord.com/channels/851854972600451112/852249738002956311/940508883095552000
> 
> > The ability to use different themes to different windows (or tabs).
> > For certain tasks I always open a different new terminal window, and I would like it to be distinct so I could (a) quickly identify it (by color), (b) remember that I'm on a session with possible non-trivial side effects / important task.

+1 for this, but I use window titles instead of different themes. I always have multiple tabs grouped in separate windows. Each window has a context, say a specific microservice I'm running locally, and inside it multiple tabs.

The lack of a window title bar makes it hard to identify windows at a glance with App Exposé, as shown below on iTerm2:

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/66829/172629962-8e1484f5-5908-429a-af55-91d8cea60631.png">

**Comment by joehannes** [comment]
I think it would be wonderful if you guys didn't invent your own profiles/session management formats/schemas, but reused the existing one's of tmux/tmuxinator where possible.

If you would integrate that formats into warp
* simulating/emulating tmux.conf as of it's statusline-components
* bringing in sessions via tmuxinator.yml

tmux guys could possibly just switch over to warp and still feel independent and first class.
* they can continue fostering their profiles/sessions in an independent format
* they can just switch to warp, and not just throw warp-sessions/profiles away ... but use the best of both worlds

You
* don't need to maintain your own format/schema
* win over the tmux guys
* can make use and even contribute to an existing OSS community, gaining popularity

**Comment by dandwal** [comment]
Just started using warp, love it. But unfortunately can't stick with it in its current state without this feature.

My problem is when using microservices and micro frontend I have many panels open and each runs an app. I need multiple panels in the same tab so I can see them talking to each other.

Unfortunately with Warp it's currently not easy to identify which panel is running which app. Something like iTerms 'badges' that sits in the background of the panel behind the main text would suffice if not wanting to introduce panel window bars.

**Reply by sdx-crispin** [reply]
Agreed, this is essential!

**Comment by EdanSaltz** [comment]
The ability to resume a session makes Tmux very nice. People who work with laptops can return to a disconnected session. This would require Warp to install software on a target machine. This functionality would allow closing a window to not end the session. There would need to be a subsystem to manage sessions.

**Reply by ild0tt0re** [reply]
Oh yes please! Native warp support for session handling like tmux does without losing the ssh session and stop the related task running on the remote machine would allow us to just use warp without adding other tools

**Comment by tukusejssirs** [comment]
Is this about workspace/project management? I wish I could save a window (all its tabs and their context + worspace name), close and re-open it anytime will its full context, just like we can do it JetBrains IDEs or Sublime Text. Currently, I have 17 windows open (~60 tabs open in total) just because of this.

If this discussion is unrelated to this, tell me, I’ll create a separate one. :pray:

**Reply by howiejc** [reply]
Yeah, this is a major reason why I don't find Warp as useful as it could be.
It's way too easy to lose information/data or useful context for no good reason.

If we could easily save (duplicate, etc.) sessions / "workspaces" / projects, Warp would be a serious contender...

**Reply by SheepDomination** [reply]
I find Warp restores sessions very well, maybe a recent update ?
